compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -0,0 +1,329 @@
+            let mut block_to_instructions = HashMap::default();
+            // Track already found loop invariants
+            let mut loop_invariants = HashSet::default();
+            for block in loop_.blocks.iter() {
+                let mut instructions_to_keep = Vec::new();
+                for instruction_id in inserter.function.dfg[*block].take_instructions() {
+                    let mut is_not_loop_invariant = false;
+                    // The list of blocks for a nested loop contain any inner loops as well.
+                    // We may have already re-inserted new instructions if two loops share blocks
+                    // so we need to map all the values in the instruction which we want to check.
+                    let (instruction, _) = inserter.map_instruction(instruction_id);
+                    instruction.for_each_value(|value| {
+                        // We are implicitly checking whether the values are constant as well.
+                        // The set of values defined in the loop only contains instruction results and block parameters
+                        // which cannot be constants.
+                        is_not_loop_invariant |=
+                            defined_in_loop.contains(&value) && !loop_invariants.contains(&value);
+                    });
+
+                    if !is_not_loop_invariant
+                        && instruction.can_be_deduplicated(&inserter.function.dfg, false)
+                    {
+                        // We need to collect the results as we then mutably borrow to resolve the ValueIds
+                        let results =
+                            inserter.function.dfg.instruction_results(instruction_id).to_vec();
+                        let results = results.into_iter().map(|value| inserter.resolve(value));
+
+                        loop_invariants.extend(results);
+                        instructions_to_hoist.push(instruction_id);
+                    } else {
+                        instructions_to_keep.push(instruction_id);
+                    }
+                }
+                block_to_instructions.insert(*block, instructions_to_keep);
+            }
+
+            // Insert instructions we wish to hoist into the pre-header first
+            // The loop body and exit are dependent upon these instructions being mapped first
+            for instruction_id in instructions_to_hoist {
+                let old_results =
+                    inserter.function.dfg.instruction_results(instruction_id).to_vec();
+
+                let (instruction, _) = inserter.map_instruction(instruction_id);
+                let new_results = push_instruction(
+                    instruction_id,
+                    instruction,
+                    &old_results,
+                    pre_header,
+                    &mut inserter.function.dfg,
+                );
+
+                replace_result_ids(&mut inserter.function.dfg, &old_results, &new_results);
```

Comment:
Do we expect `old_results` and `new_results` to always have the same number of elements? If so, please add an assertion here or in `replace_result_ids`, which currently `zip`'s the two `Vec`'s, which will silently skip results if one of the `Vec`'s is shorter than the other